src/openms/include/OpenMS/ANALYSIS/ID/BayesianProteinInferenceAlgorithm.h

```diff
@@ -0,0 +1,125 @@
+#define OPENMS_ANALYSIS_ID_BAYESIANPROTEININFERENCE_H
+
+//#define INFERENCE_BENCH
+
+#include <OpenMS/ANALYSIS/ID/MessagePasserFactory.h>
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/CONCEPT/ProgressLogger.h>
+#include <OpenMS/MATH/MISC/GridSearch.h>
+#include <vector>
+#include <functional>
+#include <boost/optional.hpp>
+
+namespace OpenMS
+{
+  class ConsensusMap;
+  class IDBoostGraph;
+  class PeptideIdentification;
+  class ProteinIdentification;
+  class ExperimentalDesign;
+
+  class OPENMS_DLLAPI BayesianProteinInferenceAlgorithm :
+      public DefaultParamHandler,
+      public ProgressLogger
+  {
+  public:
+    /// Constructor
+    explicit BayesianProteinInferenceAlgorithm(unsigned int debug_lvl = 0);
+
+    /// Destructor
+    ~BayesianProteinInferenceAlgorithm() override = default;
+
+    void updateMembers_() override;
+
+    //Note: How to perform group inference
+    // Three options:
+    // - (as implemented) use the automatically created indist. groups and report their posterior
+    // - (can be done additionally) collapse proteins to groups beforehand and run inference
+    // - (if no single protein scores wanted at all) calculate prior from proteins for the group
+    //  beforehand and remove proteins from network (saves computation
+    //  because messages are not passed from prots to groups anymore.
+
+    /// A function object to pass into the IDBoostGraph class to perform algorithms on
+    /// connected components
+    class GraphInferenceFunctor;
+
+    /// A function object to pass into the IDBoostGraph class to perform algorithms on
+    /// connected components
+    class ExtendedGraphInferenceFunctor;
+
+    /// A function object to pass into the GridSearch;
+    struct GridSearchEvaluator;
+
+    /// Perform inference. Writes its results into protein and (optionally) peptide hits (new score).
+    /// Optionally adds indistinguishable protein groups with separate scores, too.
+    /// Currently only takes first proteinID run.
+    /// TODO loop over all runs
+    void inferPosteriorProbabilities(
+        std::vector<ProteinIdentification>& proteinIDs,
+        std::vector<PeptideIdentification>& peptideIDs,
+        boost::optional<const ExperimentalDesign&> = boost::optional<const ExperimentalDesign&>());
+
+    void inferPosteriorProbabilities(
+        ConsensusMap& cmap,
+        boost::optional<const ExperimentalDesign&> = boost::optional<const ExperimentalDesign&>());
+
+  private:
+
+    void inferPosteriorProbabilities_(IDBoostGraph& ibg);
+
+    GridSearch<double,double,double> initGridSearchFromParams_(
+        std::vector<double>& alpha_search,
+        std::vector<double>& beta_search,
+        std::vector<double>& gamma_search
+        );
+
+    void setScoreTypeAndSettings_(ProteinIdentification& proteinIDs);
+
+    std::function<void(PeptideIdentification&/*, const String& run_id*/)> checkConvertAndFilterPepHits_;
```

Comment:
what about the commented out variable?

Reply:
This is a TODO to support looping over multiple runs. I documented it.